### ISSUE TITLE
Long addition error

### ISSUE THREAD

**Issue by rixner** [skulpt/skulpt#729]
There seems to be some sort of error in the long implementation:

```python
# Expected: 10821521012635269
# Skulpt: 10821521012635268
a = 8000000000000000
b = 2821521012635269
print a + b
```

I haven't had time to look into the issue, but wanted to report it so that it doesn't get lost.  I haven't seen this in other cases and I'm not sure if there's anything special about one or both of those particular numbers.

**Comment by albertjan** [comment]
I think I know what's going on here. 😄 

It's not a skulpt problem it's a javascript problem:

```
> (8000000000000000 + 2821521012635269) - 2821521012635269
< 7999999999999999
```

**Comment by j-f1** [comment]
Maybe something like [BigInteger.js](https://github.com/peterolson/BigInteger.js) would help here by letting us store large integers that JS doesn’t like.

**Comment by albertjan** [comment]
It would I think, but this is too good of a problem not to figure out 😄

**Comment by albertjan** [comment]
I'm pretty sure this is the problem:
```
> (2**53 + 1) === 2**53
< true
```

**Comment by albertjan** [comment]
We just have to think of a performant way to solve this. The idea is that if the result of the addition is larger then `2**53` we have to pass it on to biginteger.js

**Comment by rixner** [comment]
Nice catch @albertjan .  I agree, it looks like this is a question of how to modify the ```Sk.builtin.int_.prototype.nb$*``` functions to check for overflow without killing performance...

**Comment by rixner** [comment]
I have students who continually trip over this.  Since there is no longer a long type in Python 3 and BigInt is coming to Javascript, does it make sense to rewrite our int class to use BigInt (https://github.com/GoogleChromeLabs/jsbi until it is native everywhere, at which point we should be able to convert to use it relatively easily) and toss the long type?

I guess one issue is that I suspect that using this always for ints will result in slowdowns for small integer math.  And it's clear that switching back and forth appropriately is non-trivial or one of us would have fixed this bug long ago.

**Comment by rixner** [comment]
Yeah, it seemingly kills small integer performance: https://peterolson.github.io/BigInteger.js/benchmark/

That page doesn't show what the native BigInt does, though.

So, we would still need to fix the way we detect when we should be using big integers and when we should not be.

**Comment by rixner** [comment]
Never mind, the fix isn't that bad (#860).  I still think that in the long run we should move towards eliminating long and using ```BigInt```.

**Comment by albertjan** [comment]
In my es6 branch I pulled in a npm package that's a polyfill for bigint, I agree we should move to something like that, in favour of what we have now. The api is pretty similar.
